File: src/game/deck.ts
import { Card, Rank, Suit, SUITS } from "./types";

// 5-player Black Queen deck (project-specific variant):
//   • 2 copies of every: 5, 8, 9, 10, J, Q, K, A in all 4 suits  -> 2 × 8 × 4 = 64 cards
//   • 1 copy of 7♠ (the only 7 in the deck)                       -> +1
//   = 65 cards total, dealt 13 per player.
//
// Point cards: A=15, 10=10, 5=5, Q♠=30. With 2 decks that's
//   2×30 + 8×15 + 8×10 + 8×5 = 60+120+80+40 = 300 total points.

const KEPT_RANKS: Rank[] = [5, 7, 8, 9, 10, 11, 12, 13, 14];

export function buildDeck5p(): Card[] {
  const deck: Card[] = [];
  for (let copy = 0; copy < 2; copy++) {
    for (const suit of SUITS) {
      for (const rank of KEPT_RANKS) {
        if (rank === 7) continue; // 7s handled specially below
        deck.push({ suit, rank, id: `${suit}${rank}_${copy}` });
      }
    }
  }
  // Single 7♠
  deck.push({ suit: "S", rank: 7, id: "S7_0" });
  if (deck.length !== 65) throw new Error(`Expected 65 cards, got ${deck.length}`);
  return deck;
}

export function shuffle<T>(arr: T[], rng: () => number = Math.random): T[] {
  const a = arr.slice();
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

export function deal5p(deck: Card[]): Card[][] {
  if (deck.length !== 65) throw new Error("Deck must be 65 cards");
  const hands: Card[][] = [[], [], [], [], []];
  for (let i = 0; i < 65; i++) hands[i % 5].push(deck[i]);
  for (const h of hands) h.sort(handSort);
  return hands;
}

/**
 * "Light shuffle" deal: biases each hand toward a dominant suit (~7–9 cards) plus
 * a near-void in another suit. Produces more dramatic bids and louder voids
 * than a fully-random shuffle, which tends to give boring 4/3/3/3 hands.
 *
 * Algorithm: assign each player a dominant + a weak suit (rotated randomly so
 * all 4 suits are favored by at least one of the 5 players). Then walk the
 * shuffled deck and assign each card to a player by weighted random (5× weight
 * on dominant, 0.4× on weak, 1× on neutral, 0 if hand already full).
 */
export function deal5pLight(deck: Card[]): Card[][] {
  if (deck.length !== 65) throw new Error("Deck must be 65 cards");
  const suits: Suit[] = ["S", "H", "D", "C"];
  const shuffledSuits = shuffle(suits);
  // Five favored suits (one suit doubled), assigned to players in a random order.
  const dominantList: Suit[] = [shuffledSuits[0], shuffledSuits[1], shuffledSuits[2], shuffledSuits[3], shuffledSuits[Math.floor(Math.random() * 4)]];
  const playerOrder = shuffle([0, 1, 2, 3, 4]);
  const playerDominant: Record<number, Suit> = {};
  const playerWeak: Record<number, Suit> = {};
  for (let i = 0; i < 5; i++) {
    const p = playerOrder[i];
    playerDominant[p] = dominantList[i];
    const others = suits.filter((s) => s !== playerDominant[p]);
    playerWeak[p] = others[Math.floor(Math.random() * others.length)];
  }

  const hands: Card[][] = [[], [], [], [], []];
  for (const card of deck) {
    const weights: number[] = [];
    let total = 0;
    for (let i = 0; i < 5; i++) {
      let w: number;
      if (hands[i].length >= 13) w = 0;
      else if (card.suit === playerDominant[i]) w = 5;
      else if (card.suit === playerWeak[i]) w = 0.4;
      else w = 1;
      weights.push(w);
      total += w;
    }
    let r = Math.random() * total;
    let chosen = 4;
    for (let i = 0; i < 5; i++) {
      r -= weights[i];
      if (r <= 0) { chosen = i; break; }
    }
    hands[chosen].push(card);
  }
  // Sanity: any hand under 13 cards (shouldn't happen if weights sum > 0, but be safe)
  for (let i = 0; i < 5; i++) {
    if (hands[i].length !== 13) {
      // Fallback: re-balance by moving from largest hand
      while (hands[i].length < 13) {
        const biggest = hands.reduce((acc, h, idx) => h.length > acc.len ? { idx, len: h.length } : acc, { idx: 0, len: 0 });
        hands[i].push(hands[biggest.idx].pop()!);
      }
      while (hands[i].length > 13) {
        const smallest = hands.reduce((acc, h, idx) => (idx !== i && h.length < acc.len) ? { idx, len: h.length } : acc, { idx: -1, len: 14 });
        if (smallest.idx < 0) break;
        hands[smallest.idx].push(hands[i].pop()!);
      }
    }
  }
  for (const h of hands) h.sort(handSort);
  return hands;
}

export function handSort(a: Card, b: Card): number {
  const order: Record<string, number> = { S: 0, H: 1, C: 2, D: 3 };
  if (a.suit !== b.suit) return order[a.suit] - order[b.suit];
  return b.rank - a.rank;
}

File: rust/crates/bq-engine/src/deck.rs
use crate::rng::GameRng;
use crate::types::{Card, Suit};
use rand::seq::SliceRandom;

/// 65-card Black Queen deck: two standard decks with 2/3/4/6/7 removed,
/// then exactly one 7♠ added back. Verified at build:
///   - 4 suits × 8 ranks {5,8,9,10,J,Q,K,A} × 2 copies = 64
///   - + 1×7♠ = 65 cards, 300 total points.
pub fn build_deck() -> Vec<Card> {
    let ranks_full: [u8; 8] = [5, 8, 9, 10, 11, 12, 13, 14];
    let mut deck = Vec::with_capacity(65);
    for _ in 0..2 {
        for &suit in &Suit::ALL {
            for &rank in &ranks_full {
                deck.push(Card { suit, rank });
            }
        }
    }
    deck.push(Card { suit: Suit::S, rank: 7 });
    debug_assert_eq!(deck.len(), 65);
    deck
}

pub fn shuffle_seeded(deck: &mut [Card], rng: &mut GameRng) {
    deck.shuffle(rng);
}

/// Deal in round-robin order. Each of 5 players gets 13 cards.
pub fn deal_hands(deck: Vec<Card>) -> Vec<Vec<Card>> {
    assert_eq!(deck.len(), 65);
    let mut hands: Vec<Vec<Card>> = (0..5).map(|_| Vec::with_capacity(13)).collect();
    for (i, card) in deck.into_iter().enumerate() {
        hands[i % 5].push(card);
    }
    hands
}

/// Parameterized deal — interpolates between "light" (heavily biased hands with
/// a dominant suit + near-void) and "full" (uniformly random).
///   intensity = 0 → light: dominant×5, weak×0.4, neutral×1
///   intensity = 1 → full:  uniform — same distribution as `deal_hands`
/// Linear blend in between.
///
/// Mirrors `deal5pLight` in src/game/deck.ts. The shuffled `deck` argument
/// provides the iteration order; assignment to each player is weighted-random
/// using the supplied RNG. Always returns 13 cards per player.
pub fn deal_hands_light(deck: Vec<Card>, intensity: f64, rng: &mut GameRng) -> Vec<Vec<Card>> {
    use rand::Rng;
    assert_eq!(deck.len(), 65);
    let t = intensity.clamp(0.0, 1.0);
    // Linear blend between light extremes (5, 0.4) and uniform (1, 1).
    let dominant_w = 5.0 + (1.0 - 5.0) * t;
    let weak_w = 0.4 + (1.0 - 0.4) * t;
    // (neutral weight is always 1.0)

    // Five favored suits, one doubled; assigned to players in a random order.
    let mut suits = Suit::ALL.to_vec();
    suits.shuffle(rng);
    let extra_idx: usize = rng.gen_range(0..4);
    let dominant_list: [Suit; 5] = [suits[0], suits[1], suits[2], suits[3], suits[extra_idx]];
    let mut player_order: [usize; 5] = [0, 1, 2, 3, 4];
    player_order.shuffle(rng);
    let mut player_dominant: [Suit; 5] = [Suit::S; 5];
    let mut player_weak: [Suit; 5] = [Suit::S; 5];
    for i in 0..5 {
        let p = player_order[i];
        player_dominant[p] = dominant_list[i];
        let others: Vec<Suit> = Suit::ALL.iter().copied()
            .filter(|s| *s != player_dominant[p]).collect();
        let pick: usize = rng.gen_range(0..others.len());
        player_weak[p] = others[pick];
    }

    let mut hands: Vec<Vec<Card>> = (0..5).map(|_| Vec::with_capacity(13)).collect();
    for card in deck {
        let mut weights = [0.0f64; 5];
        let mut total = 0.0;
        for i in 0..5 {
            let w = if hands[i].len() >= 13 { 0.0 }
                else if card.suit == player_dominant[i] { dominant_w }
                else if card.suit == player_weak[i] { weak_w }
                else { 1.0 };
            weights[i] = w;
            total += w;
        }
        let mut r: f64 = rng.gen_range(0.0..total.max(f64::MIN_POSITIVE));
        let mut chosen = 4usize;
        for i in 0..5 {
            r -= weights[i];
            if r <= 0.0 { chosen = i; break; }
        }
        hands[chosen].push(card);
    }

    // Safety re-balance: any hand under 13 borrows from the largest.
    for i in 0..5 {
        while hands[i].len() < 13 {
            let (biggest_idx, biggest_len) = hands.iter().enumerate()
                .map(|(idx, h)| (idx, h.len()))
                .max_by_key(|&(_, len)| len)
                .unwrap_or((0, 0));
            if biggest_len <= hands[i].len() { break; }
            let card = hands[biggest_idx].pop().unwrap();
            hands[i].push(card);
        }
        while hands[i].len() > 13 {
            let smallest = hands.iter().enumerate()
                .filter(|(idx, h)| *idx != i && h.len() < 13)
                .map(|(idx, h)| (idx, h.len()))
                .min_by_key(|&(_, len)| len);
            match smallest {
                Some((idx, _)) => {
                    let card = hands[i].pop().unwrap();
                    hands[idx].push(card);
                }
                None => break,
            }
        }
    }
    hands
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::rng::from_seed;

    #[test]
    fn deck_has_65_cards() {
        assert_eq!(build_deck().len(), 65);
    }

    #[test]
    fn deck_total_points_is_300() {
        let total: u16 = build_deck().iter().map(|c| c.points()).sum();
        assert_eq!(total, 300);
    }

    #[test]
    fn exactly_one_seven_of_spades_no_other_sevens() {
        let deck = build_deck();
        let sevens: Vec<_> = deck.iter().filter(|c| c.rank == 7).collect();
        assert_eq!(sevens.len(), 1);
        assert_eq!(sevens[0].suit, Suit::S);
    }

    #[test]
    fn no_low_ranks_in_deck() {
        let deck = build_deck();
        for c in &deck {
            assert!(!matches!(c.rank, 2 | 3 | 4 | 6), "stray low rank in deck: {:?}", c);
        }
    }

    #[test]
    fn shuffle_is_deterministic_per_seed() {
        let mut a = build_deck();
        let mut b = build_deck();
        shuffle_seeded(&mut a, &mut from_seed(42));
        shuffle_seeded(&mut b, &mut from_seed(42));
        assert_eq!(a, b);
    }

    #[test]
    fn shuffle_differs_across_seeds() {
        let mut a = build_deck();
        let mut b = build_deck();
        shuffle_seeded(&mut a, &mut from_seed(1));
        shuffle_seeded(&mut b, &mut from_seed(2));
        assert_ne!(a, b);
    }

    #[test]
    fn deal_gives_each_player_13() {
        let hands = deal_hands(build_deck());
        assert_eq!(hands.len(), 5);
        for h in &hands {
            assert_eq!(h.len(), 13);
        }
    }
}
